# Lập trình cho Khoa học dữ liệu: Đồ án cuối kì
## Nhóm 8: Thông tin sinh viên
- #### Trello: https://trello.com/b/Tb6Sjg8O


| MSSV     | Họ tên            | GitHub                         |    Email         
|--------------|---------------------|------------------------------ |--------------------------------------------------|
| 19127009 | Lữ Thế Vỹ | [luthevy](https://github.com/luthevy) | 19127009@student.hcmus.edu.vn
| 19127416 | Mạc Văn Hưng     | [machung2001](https://github.com/machung2001)        |19127416@student.hcmus.edu.vn


# Mục lục
- #### [Lời nói đầu](#Lời-nói-đầu)
    - [Giới thiệu](#Giới-thiệu)
    - [Mục tiêu](#Mục-tiêu)
    - [Nguồn tài liệu](#Nguồn-tài-liệu)
- #### [Bước chuẩn bị](#Bước-chuẩn-bị)
    - [Nhận diện môi trường code](#Môi-trường-code)
    - [Import thư viện](#Import-các-thư-viện-cần-thiết)
- #### [Khám phá dữ liệu](#Thu-thập-dữ-liệu)
    - [Đọc dữ liệu](#Đọc-dữ-liệu-từ-file-csv-vào-dataframe)
    - [Thông tin cơ bản](#Thông-tin-cơ-bản)
        - [Số dòng, số cột của tập dữ liệu](#Dữ-liệu-gồm-có-bao-nhiêu-dòng-và-bao-nhiêu-cột?)
        - [Ý nghĩa từng cột](#Mỗi-cột-có-ý-nghĩa-gì?)
        - [Kiểu dữ liệu từng cột](#Mỗi-cột-hiện-đang-có-kiểu-dữ-liệu-gì?)
    - [Tiền xử lý](#Tiền-xử-lý)
        - [Cột cần loại bỏ](#Cột-cần-loại-bỏ)
        - [Cột bất thường](#Cột-bất-thường)
        - [Dòng bất thường](#Dòng-bất-thường)
        - [Dòng bị lặp](#Dòng-bị-lặp)
- #### [Khám phá dữ liệu (tiếp theo)](#Khám-phá-dữ-liệu-(tiếp-theo))
    - [Phân bố dữ liệu kiểu numerical](#Phân-bố-dữ-liệu-kiểu-numerical)
    - [Phân bố dữ liệu kiểu categorical](#Phân-bố-dữ-liệu-kiểu-categorical)
- #### [Đặt câu hỏi](#Đặt-câu-hỏi)


_____________________________________________

## Lời nói đầu

### Giới thiệu

### Mục tiêu

### Nguồn tài liệu

______

## Bước chuẩn bị

___

### Môi trường code

In [1]:
import sys
sys.executable

'F:\\Anaconda\\python.exe'

### Import các thư viện cần thiết

In [2]:
# Import thư viện cần dùng
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

---

## Khám phá dữ liệu

---

### Đọc dữ liệu từ file csv vào dataframe

In [3]:
hcs_df = pd.read_csv('data/healthcare-dataset-stroke-data.csv')
hcs_df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

### Thông tin cơ bản

#### Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?

In [4]:
num_rows = hcs_df.shape[0]
num_cols = hcs_df.shape[1]

hcs_df.shape

(5110, 12)

Dữ liệu có 12 cột và 5110 dòng.

#### Mỗi cột có ý nghĩa gì?

In [5]:
with open('data/description.txt', 'r') as f:
    print(f.read())

1) id:                 unique identifier
2) gender:             "Male", "Female" or "Other"
3) age:                age of the patient
4) hypertension:       0 if the patient doesn't have hypertension, 1 if the patient has hypertension
5) heart_disease:      0 if the patient doesn't have any heart diseases, 1 if the patient has a heart disease
6) ever_married:       "No" or "Yes"
7) work_type:          "children", "Govt_jov", "Never_worked", "Private" or "Self-employed"
8) Residence_type:     "Rural" or "Urban"
9) avg_glucose_level:  average glucose level in blood
10) bmi:               body mass index
11) smoking_status:    "formerly smoked", "never smoked", "smokes" or "Unknown"*
12) stroke.            1 if the patient had a stroke or 0 if not
*Note: "Unknown" in smoking_status means that the information is unavailable for this patient


- **id**: mã id
- **gender**: giới tính ("Male", Female" và "Other")
- **age**: tuổi tác
- **hypertension**: bị huyết áp cao (0 là không, 1 là có)
- **heart_disease**: bị bệnh tim (0 là không, 1 là có)
- **ever_married**: tình trạng kết hôn ("No" và "Yes")
- **work_type**: loại nghề nghiệp ("children", "Govt_jov", "Never_worked", "Private" và "Self-employed")
- **Residence_type**: loại cư dân ("Rural" hoặc "Urban")
- **avg_glucose_level**: lượng đường trung bình trong máu
- **bmi**: chỉ số khối cơ thể
- **smoking_status**: tình trạng hút thuốc ("formerly smoked", "never smoked", "smokes" và "Unknown")
- **stroke**: bị đột quỵ (0 là không, 1 là có)

#### Mỗi cột hiện đang có kiểu dữ liệu gì?

In [6]:
hcs_df.dtypes

id                     int64
gender                object
age                  float64
hypertension           int64
heart_disease          int64
ever_married          object
work_type             object
Residence_type        object
avg_glucose_level    float64
bmi                  float64
smoking_status        object
stroke                 int64
dtype: object

_______

### Tiền xử lý

#### Cột cần loại bỏ
Trong tập dữ liệu `healthcare-dataset-stroke-data.csv`, có tất cả 12 cột dữ liệu chứa các thông tin liên quan đến sức khoẻ. Duy chỉ có một cột và là cột đầu tiên `id` chứa thông tin mã số phân biệt của mỗi đối tượng tham gia khảo sát. Cột `id` nhìn chung không có giá trị hữu ích để dùng cho việc đặt câu hỏi sau này (do giá trị của nó không liên quan đến chủ đề sức khoẻ), nên ta sẽ loại bỏ nó.

In [7]:
hcs_df.drop(columns=['id'], inplace=True)
hcs_df.head()

gender   age  hypertension  heart_disease ever_married      work_type  \
0    Male  67.0             0              1          Yes        Private   
1  Female  61.0             0              0          Yes  Self-employed   
2    Male  80.0             0              1          Yes        Private   
3  Female  49.0             0              0          Yes        Private   
4  Female  79.0             1              0          Yes  Self-employed   

  Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
0          Urban             228.69  36.6  formerly smoked       1  
1          Rural             202.21   NaN     never smoked       1  
2          Rural             105.92  32.5     never smoked       1  
3          Urban             171.23  34.4           smokes       1  
4          Rural             174.12  24.0     never smoked       1

#### Cột bất thường
Dựa theo thông tin về kiểu dữ liệu của từng cột, đối chiếu với dữ liệu trên bảng, ta thấy dữ liệu từng cột không có gì khác thường.


In [13]:
hcs_df.dtypes

gender                object
age                  float64
hypertension           int64
heart_disease          int64
ever_married          object
work_type             object
Residence_type        object
avg_glucose_level    float64
bmi                  float64
smoking_status        object
stroke                 int64
dtype: object

In [11]:
hcs_df.head()

gender   age  hypertension  heart_disease ever_married      work_type  \
0    Male  67.0             0              1          Yes        Private   
1  Female  61.0             0              0          Yes  Self-employed   
2    Male  80.0             0              1          Yes        Private   
3  Female  49.0             0              0          Yes        Private   
4  Female  79.0             1              0          Yes  Self-employed   

  Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
0          Urban             228.69  36.6  formerly smoked       1  
1          Rural             202.21   NaN     never smoked       1  
2          Rural             105.92  32.5     never smoked       1  
3          Urban             171.23  34.4           smokes       1  
4          Rural             174.12  24.0     never smoked       1

#### Dòng bất thường
Mỗi dòng liệt kê các thông tin sức khoẻ của người tham gia khảo sát, ứng với từng cột. Các dòng dữ liệu nhìn chung không có gì bất thường, trừ các dòng bị thiếu hụt dữ liệu (ở cột `bmi`). Có 201 ô dữ liệu thiếu ở cột `bmi`, tương ứng 201 dòng bị thiếu dữ liệu.

In [14]:
hcs_df.isnull().sum()

gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

In [25]:
hcs_df['bmi'] = hcs_df['bmi'].fillna(hcs_df['bmi'].median())

In [24]:
hcs_df[hcs_df['bmi'].isnull()]

gender   age  hypertension  heart_disease ever_married      work_type  \
1     Female  61.0             0              0          Yes  Self-employed   
8     Female  59.0             0              0          Yes        Private   
13      Male  78.0             0              1          Yes        Private   
19      Male  57.0             0              1           No       Govt_job   
27      Male  58.0             0              0          Yes        Private   
...      ...   ...           ...            ...          ...            ...   
5039    Male  41.0             0              0           No        Private   
5048    Male  40.0             0              0          Yes        Private   
5093  Female  45.0             1              0          Yes       Govt_job   
5099    Male  40.0             0              0          Yes        Private   
5105  Female  80.0             1              0          Yes        Private   

     Residence_type  avg_glucose_level  bmi   smoking_status  stroke  
1             Rural             202.21  NaN     never smoked       1  
8             Rural              76.15  NaN          Unknown       1  
13            Urban             219.84  NaN          Unknown       1  
19            Urban             217.08  NaN          Unknown       1  
27            Rural             189.84  NaN          Unknown       1  
...             ...                ...  ...              ...     ...  
5039          Rural              70.15  NaN  formerly smoked       0  
5048          Urban             191.15  NaN           smokes       0  
5093          Rural              95.02  NaN           smokes       0  
5099          Rural              83.94  NaN           smokes       0  
5105          Urban              83.75  NaN     never smoked       0  

[201 rows x 11 columns]

#### Dòng bị lặp
Dữ liệu từng dòng của dataset `healthcare-dataset-stroke-data.csv` có bị trùng lặp nhau không?

In [22]:
hcs_total_dupRows = np.sum(hcs_df.duplicated())

pd.DataFrame({'Dataset': ['Healthcare'],
             'Duplicate count': [hcs_total_dupRows]})


Dataset  Duplicate count
0  Healthcare                0

In [ ]:
    - [Kiểu dữ liệu từng cột](#Mỗi-cột-hiện-đang-có-kiểu-dữ-liệu-gì?)
    - [Nhận xét](#Nhận-xét)
        - [Cột cần loại bỏ](#Cột-cần-loại-bỏ)
        - [Cột bất thường](#Cột-bất-thường)
        - [Dòng bất thường](#Dòng-bất-thường)
        - [Dòng bị lặp](#Dòng-bị-lặp)
- #### [Khám phá dữ liệu (tiếp theo)](#Khám-phá-dữ-liệu-(tiếp theo))
    - [Phân bố dữ liệu kiểu numerical](#Phân-bố-dữ-liệu-kiểu-numerical)
    - [Phân bố dữ liệu kiểu categorical](#Phân-bố-dữ-liệu-kiểu-categorical)
- #### [Đặt câu hỏi](#Đặt-câu-hỏi)

### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?

Hiện có 6 cột thuộc dạng numeric: "age", "hypertension", "heart_disease", "avg_glucose_level", "bmi", "stroke". Nhưng 3 cột "hypertension", "heart_disease", "stroke" do mã được đánh dưới dạng số nên 3 cột này sẽ là dạng categorical. Vì vậy, chỉ có 3 cột numeric là "age", "avg_glucose_level", "bmi". 

In [27]:
def missing_ratio(col):
    return col.isna().mean() * 100

In [28]:
num_col_profiles_df = hcs_df[["age", "avg_glucose_level", "bmi"]]
num_col_profiles_df = num_col_profiles_df.agg([missing_ratio, pd.Series.min, pd.Series.max])

### Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?

Ban đầu sẽ có 5 cột thuộc dạng categorical: "gender", "ever_married", "work_type", "Residence_type", "smoking_status".
Bởi vì lý do ở trên nên sẽ có thêm 3 cột: "hypertension", "heart_disease", "stroke". Vì vậy, tổng cộng có 8 cột thuộc dạng categorical: "id", "gender", "hypertension", "heart_disease","work_type", "Residence_type", "smoking_status", "stroke".

In [29]:
def unique_non_null(s):
    return s.dropna().unique()

In [30]:
cate_col_profiles_df = hcs_df[["gender", "hypertension", "heart_disease","work_type", 
                               "Residence_type", "smoking_status", "stroke"]]
cate_col_profiles_df = cate_col_profiles_df.agg([missing_ratio, unique_non_null, pd.Series.nunique])
cate_col_profiles_df = cate_col_profiles_df.rename(index={'unique_non_null': 'diff_vals', 'nunique': 'num_diff_vals'})